pkg/util/config/feature_gate.go

```diff
@@ -37,6 +37,8 @@ const (
 	// values override the default set by allAlpha.
 	// e.g. allAlpha=false,newFeature=true will result in newFeature=true
 	allAlpha = "allAlpha"
+
+	externalTrafficLocalOnly = "externalTrafficLocalOnly"
```

Comment:
yes, you should check AllAlpha.

@jlowdermilk what is the intended logic here?  There's no way to determine
if a specific flag was explicitly disabled or if it is default disabled.
For example: if I set `allAlpha=true`, it is pretty clear that I mean to
enable `externalTrafficLocalOnly`, too.  If I
set `allAlpha=true,externalTrafficLocalOnly=false`, should that disable
`externalTrafficLocalOnly`?  If so, how do I tell the difference between
the default (false) and explicit false?  Or should the behavior simply be
"OR"?

options:

a) make all of these flags tri-state (*bool) to indicate true, false, don't
care: (complex)

b) make the intersection of "allAlpha" and specific gates be logical OR:
(non-obvious what `allAlpha=true,myFlag=false` does; all callers have to
check both)

c) special-case allAlpha (and allBeta?) to literally iterate all gates
designated "ALPHA" and settheir value, and then allow individual gates to
change it (most obvious, most work)  I will see if a PR to do this is easy
enough


wrt naming, I don't think we have come up with anything better than what we
had, so let's KISS.

wrt e2e: Can you ping test gang to see their thoughts on e2e testing alphas
by default?  Do we gate-enable all alpha flags for e2e or not test alphas,
or do a different jenkins?



On Mon, Aug 22, 2016 at 10:13 AM, Girish Kalele <notifications@github.com>
wrote:

> In pkg/util/config/feature_gate.go
> <https://github.com/kubernetes/kubernetes/pull/29409#discussion_r75717897>
> :
>
> > @@ -37,6 +37,8 @@ const (
> >  	// values override the default set by allAlpha.
> >  	// e.g. allAlpha=false,newFeature=true will result in newFeature=true
> >  	allAlpha = "allAlpha"
> > +
> > +	externalTrafficLocalOnly = "externalTrafficLocalOnly"
>
> Also, should I also check the above "AllAlpha" gate ? It seems to imply
> all Alpha features enabled.
>
> How about prefixing with Allow/Block ?
>
> For 1.4, "AllowExtTrafficLocalEndpointAnnotation" ?
>
> And in the future, negative behaviour, "BlockExtTrafficLocalEndpointAnnotation"
> ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/29409/files/59883dc63dbea4bc31ef2ef85d3274a0ea383e78..9875b3b6f8892cf2c1e4c0aa2009b5161448c094#r75717897>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVG_Ctj5xVBczrGznP1TLYpkSk191ks5qidiegaJpZM4JSUXH>
> .
>